File: json/json-annotation-processor/src/main/java/ru/tinkoff/kora/json/annotation/processor/writer/JsonWriterGenerator.java
package ru.tinkoff.kora.json.annotation.processor.writer;

import com.squareup.javapoet.*;
import ru.tinkoff.kora.annotation.processor.common.CommonClassNames;
import ru.tinkoff.kora.annotation.processor.common.CommonUtils;
import ru.tinkoff.kora.json.annotation.processor.JsonTypes;
import ru.tinkoff.kora.json.annotation.processor.JsonUtils;
import ru.tinkoff.kora.json.annotation.processor.KnownType;
import ru.tinkoff.kora.json.annotation.processor.writer.JsonClassWriterMeta.FieldMeta;

import javax.annotation.Nullable;
import javax.annotation.processing.ProcessingEnvironment;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.util.Types;
import java.io.IOException;

public class JsonWriterGenerator {
    private final Types types;

    public JsonWriterGenerator(ProcessingEnvironment processingEnvironment) {
        this.types = processingEnvironment.getTypeUtils();
    }

    @Nullable
    public TypeSpec generate(JsonClassWriterMeta meta) {

        var typeBuilder = TypeSpec.classBuilder(JsonUtils.jsonWriterName(meta.typeElement()))
            .addAnnotation(AnnotationSpec.builder(CommonClassNames.koraGenerated)
                .addMember("value", CodeBlock.of("$S", JsonWriterGenerator.class.getCanonicalName()))
                .build())
            .addSuperinterface(ParameterizedTypeName.get(JsonTypes.jsonWriter, TypeName.get(meta.typeElement().asType())))
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addOriginatingElement(meta.typeElement());

        for (var typeParameter : meta.typeElement().getTypeParameters()) {
            typeBuilder.addTypeVariable(TypeVariableName.get(typeParameter));
        }


        this.addWriters(typeBuilder, meta);
        for (var field : meta.fields()) {
            typeBuilder.addField(FieldSpec.builder(JsonTypes.serializedString, this.jsonNameStaticName(field), Modifier.PRIVATE, Modifier.STATIC, Modifier.FINAL)
                .initializer(CodeBlock.of("new $T($S)", JsonTypes.serializedString, field.jsonName()))
                .build());
        }

        var method = MethodSpec.methodBuilder("write")
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addException(IOException.class)
            .addParameter(JsonTypes.jsonGenerator, "_gen")
            .addParameter(ParameterSpec.builder(TypeName.get(meta.typeMirror()), "_object").addAnnotation(Nullable.class).build())
            .addAnnotation(Override.class);
        method.addCode("if (_object == null) {$>\n_gen.writeNull();\nreturn;$<\n}\n");
        method.addStatement("_gen.writeStartObject(_object)");

        var discriminatorField = JsonUtils.discriminatorField(types, meta.typeElement());
        if (discriminatorField != null) {
            if (meta.fields().stream().noneMatch(f -> f.jsonName().equals(discriminatorField))) {
                var discriminatorFieldValues = JsonUtils.discriminatorValue(meta.typeElement());
                method.addCode("_gen.writeFieldName($S);\n", discriminatorField);
                method.addStatement("_gen.writeString($S);", discriminatorFieldValues.get(0));
            }
        }
        for (var field : meta.fields()) {
            this.addWriteParam(method, field);
        }
        method.addStatement("_gen.writeEndObject()");

        typeBuilder.addMethod(method.build());
        return typeBuilder.build();
    }


    private void addWriters(TypeSpec.Builder typeBuilder, JsonClassWriterMeta classMeta) {
        var constructor = MethodSpec.constructorBuilder()
            .addModifiers(Modifier.PUBLIC);
        for (var field : classMeta.fields()) {
            if (field.writer() != null) {
                var fieldName = this.writerFieldName(field);
                var fieldType = TypeName.get(field.writer());
                var writerField = FieldSpec.builder(fieldType, fieldName, Modifier.PRIVATE, Modifier.FINAL);
                var writerElement = (TypeElement) this.types.asElement(field.writer());
                if (writerElement.getModifiers().contains(Modifier.FINAL)) {
                    var constructors = writerElement.getEnclosedElements().stream()
                        .filter(e -> e.getKind() == ElementKind.CONSTRUCTOR)
                        .toList();
                    if (constructors.size() == 1) {
                        writerField.addModifiers(Modifier.STATIC);
                        writerField.initializer("new $T()", field.writer());
                        typeBuilder.addField(writerField.build());
                        continue;
                    }
                }

                typeBuilder.addField(writerField.build());
                constructor.addParameter(fieldType, fieldName);
                constructor.addStatement("this.$L = $L", fieldName, fieldName);
            } else if (field.writerTypeMeta() instanceof WriterFieldType.UnknownWriterFieldType) {
                var fieldName = this.writerFieldName(field);
                var fieldType = ParameterizedTypeName.get(JsonTypes.jsonWriter, TypeName.get(field.typeMirror()));
                var writerField = FieldSpec.builder(fieldType, fieldName, Modifier.PRIVATE, Modifier.FINAL);

                typeBuilder.addField(writerField.build());
                constructor.addParameter(fieldType, fieldName);
                constructor.addStatement("this.$L = $L", fieldName, fieldName);
            }
        }
        typeBuilder.addMethod(constructor.build());
    }

    private String writerFieldName(JsonClassWriterMeta.FieldMeta field) {
        return field.accessor().getSimpleName() + "Writer";
    }

    private void addWriteParam(MethodSpec.Builder method, FieldMeta field) {
        if (field.typeMirror().getKind().isPrimitive()) {
            method.addCode("_gen.writeFieldName($L);\n", this.jsonNameStaticName(field));
            if (field.writer() == null && field.writerTypeMeta() instanceof WriterFieldType.KnownWriterFieldType typeMeta) {
                method.addCode(this.writeKnownType(typeMeta.knownType(), CodeBlock.of("_object.$L", field.accessor())));
            } else {
                method.addStatement("$L.write(_gen, _object.$L)", this.writerFieldName(field), field.accessor());
            }
            return;
        }
        var isEmptyCheck = field.includeType() == JsonClassWriterMeta.IncludeType.NON_EMPTY
            && (CommonUtils.isCollection(field.typeMirror()) || CommonUtils.isMap(field.typeMirror()));
        if (isEmptyCheck) {
            method.addCode("if (_object.$L != null && !_object.$L.isEmpty()) {$>\n", field.accessor(), field.accessor());
        } else if (field.includeType() != JsonClassWriterMeta.IncludeType.ALWAYS) {
            method.addCode("if (_object.$L != null) {$>\n", field.accessor());
        }

        method.addCode("_gen.writeFieldName($L);\n", this.jsonNameStaticName(field));
        if (field.writer() == null && field.writerTypeMeta() instanceof WriterFieldType.KnownWriterFieldType typeMeta) {
            if (field.includeType() == JsonClassWriterMeta.IncludeType.ALWAYS) {
                method.beginControlFlow("if (_object.$L == null)", field.accessor());
                method.addCode("_gen.writeNull();\n");
                method.nextControlFlow("else");
                method.addCode(this.writeKnownType(typeMeta.knownType(), CodeBlock.of("_object.$L", field.accessor())));
                method.endControlFlow();
            } else {
                method.addCode(this.writeKnownType(typeMeta.knownType(), CodeBlock.of("_object.$L", field.accessor())));
            }
        } else {
            method.addStatement("$L.write(_gen, _object.$L)", this.writerFieldName(field), field.accessor());
        }
        if (field.includeType() != JsonClassWriterMeta.IncludeType.ALWAYS) {
            method.addCode("$<}\n");
        }
    }

    private String jsonNameStaticName(FieldMeta field) {
        return "_" + field.field().getSimpleName().toString() + "_optimized_field_name";
    }

    private CodeBlock writeKnownType(KnownType.KnownTypesEnum knownType, CodeBlock value) {
        return switch (knownType) {
            case STRING -> CodeBlock.of("_gen.writeString($L);\n", value);
            case BOOLEAN_OBJECT, BOOLEAN_PRIMITIVE -> CodeBlock.of("_gen.writeBoolean($L);\n", value);
            case INTEGER_OBJECT, BIG_INTEGER, BIG_DECIMAL, DOUBLE_OBJECT, FLOAT_OBJECT, LONG_OBJECT, SHORT_OBJECT,
                INTEGER_PRIMITIVE, DOUBLE_PRIMITIVE, FLOAT_PRIMITIVE, LONG_PRIMITIVE, SHORT_PRIMITIVE -> CodeBlock.of("_gen.writeNumber($L);\n", value);
            case BINARY -> CodeBlock.of("_gen.writeBinary($L);\n", value);
            case UUID -> CodeBlock.of("_gen.writeString($L.toString());\n", value);
        };
    }
}

File: json/json-symbol-processor/src/main/kotlin/ru/tinkoff/kora/json/ksp/writer/JsonWriterGenerator.kt
package ru.tinkoff.kora.json.ksp.writer

import com.google.devtools.ksp.getConstructors
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.*
import ru.tinkoff.kora.json.ksp.JsonTypes
import ru.tinkoff.kora.json.ksp.KnownType.KnownTypesEnum
import ru.tinkoff.kora.json.ksp.KnownType.KnownTypesEnum.*
import ru.tinkoff.kora.json.ksp.discriminatorField
import ru.tinkoff.kora.json.ksp.discriminatorValues
import ru.tinkoff.kora.json.ksp.jsonWriterName
import ru.tinkoff.kora.ksp.common.KotlinPoetUtils.controlFlow
import ru.tinkoff.kora.ksp.common.KspCommonUtils.generated
import ru.tinkoff.kora.ksp.common.KspCommonUtils.toTypeName

class JsonWriterGenerator(
    private val resolver: Resolver
) {

    fun generate(meta: JsonClassWriterMeta): TypeSpec {
        val jsonClassDeclaration = meta.classDeclaration
        val typeParameterResolver = jsonClassDeclaration.typeParameters.toTypeParameterResolver()
        val typeName = meta.classDeclaration.toTypeName()
        val writerInterface = JsonTypes.jsonWriter.parameterizedBy(typeName)
        val typeBuilder = TypeSpec.classBuilder(meta.classDeclaration.jsonWriterName())
            .generated(JsonWriterGenerator::class)
        jsonClassDeclaration.containingFile?.let { typeBuilder.addOriginatingKSFile(it) }
        typeBuilder.addSuperinterface(writerInterface)

        meta.classDeclaration.typeParameters.forEach {
            typeBuilder.addTypeVariable(it.toTypeVariableName())
        }

        this.addWriters(typeBuilder, meta, typeParameterResolver)
        for (field in meta.fields) {
            typeBuilder.addProperty(
                PropertySpec.builder(this.jsonNameStaticName(field), JsonTypes.serializedString, KModifier.PRIVATE)
                    .initializer(CodeBlock.of("%T(%S)", JsonTypes.serializedString, field.jsonName))
                    .build()
            )
        }
        val functionBody = CodeBlock.builder()
        functionBody.controlFlow("if (_object == null)") {
            addStatement("_gen.writeNull()")
            addStatement("return")
        }
        functionBody.addStatement("_gen.writeStartObject(_object)")

        val discriminatorField = meta.classDeclaration.discriminatorField()
        if (discriminatorField != null) {
            if (meta.fields.none { it.jsonName == discriminatorField }) {
                val discriminatorValue = meta.classDeclaration.discriminatorValues().first()
                functionBody.addStatement("_gen.writeFieldName(%S)", discriminatorField)
                functionBody.addStatement("_gen.writeString(%S)", discriminatorValue)
            }
        }
        for (field in meta.fields) {
            this.addWriteParam(functionBody, field)
        }
        functionBody.addStatement("_gen.writeEndObject()")

        typeBuilder.addFunction(
            FunSpec.builder("write")
                .addModifiers(KModifier.PUBLIC, KModifier.OVERRIDE)
                .addParameter("_gen", JsonTypes.jsonGenerator)
                .addParameter("_object", typeName.copy(nullable = true))
                .addCode(functionBody.build())
                .build()
        )

        return typeBuilder.build()
    }

    private fun addWriters(typeBuilder: TypeSpec.Builder, classMeta: JsonClassWriterMeta, typeParameterResolver: TypeParameterResolver) {
        val constructor = FunSpec.constructorBuilder()
        for (field in classMeta.fields) {
            if (field.writer != null) {
                val fieldName: String = this.writerFieldName(field)
                val fieldType = field.writer.toTypeName(typeParameterResolver)
                val writerProp = PropertySpec.builder(fieldName, fieldType, KModifier.PRIVATE)
                val writerDeclaration = field.writer.declaration as KSClassDeclaration
                if (!writerDeclaration.modifiers.contains(com.google.devtools.ksp.symbol.Modifier.OPEN)) {
                    val constructors = writerDeclaration.getConstructors().toList()
                    if (constructors.size == 1) {
                        writerProp.initializer("%T()", field.writer.toTypeName(typeParameterResolver))
                        typeBuilder.addProperty(writerProp.build())
                        continue
                    }
                }
                typeBuilder.addProperty(writerProp.build())
                constructor.addParameter(fieldName, fieldType)
                constructor.addStatement("this.%L = %L", fieldName, fieldName)
            } else if (field.typeMeta is WriterFieldType.UnknownWriterFieldType) {
                val fieldName: String = this.writerFieldName(field)
                val fieldType = JsonTypes.jsonWriter.parameterizedBy(
                    field.typeMeta.type.toTypeName(typeParameterResolver).copy(nullable = false)
                )
                val writerField = PropertySpec.builder(fieldName, fieldType, KModifier.PRIVATE)
                typeBuilder.addProperty(writerField.build())
                constructor.addParameter(fieldName, fieldType)
                constructor.addStatement("this.%L = %L", fieldName, fieldName)
            }
        }
        typeBuilder.primaryConstructor(constructor.build())
    }

    private fun writerFieldName(field: JsonClassWriterMeta.FieldMeta): String {
        return field.accessor + "Writer"
    }

    private fun addWriteParam(function: CodeBlock.Builder, field: JsonClassWriterMeta.FieldMeta) {
        function.controlFlow("_object.%N%L.let {", field.accessor, if (field.type.isMarkedNullable) "?" else "") {
            function.add("_gen.writeFieldName(%L)\n", jsonNameStaticName(field))
            if (field.writer == null && field.typeMeta is WriterFieldType.KnownWriterFieldType) {
                function.add(writeKnownType(field.typeMeta.knownType))
            } else {
                function.addStatement("%L.write(_gen, it)", writerFieldName(field))
            }
        }
    }

    private fun jsonNameStaticName(field: JsonClassWriterMeta.FieldMeta): String {
        return "_" + field.fieldSimpleName.asString() + "_optimized_field_name"
    }

    private fun writeKnownType(knownType: KnownTypesEnum) = when (knownType) {
        KnownTypesEnum.STRING -> CodeBlock.of("_gen.writeString(it)\n")
        KnownTypesEnum.BOOLEAN -> CodeBlock.of("_gen.writeBoolean(it)\n")
        INTEGER, BIG_INTEGER, BIG_DECIMAL, KnownTypesEnum.DOUBLE, KnownTypesEnum.FLOAT, KnownTypesEnum.LONG, KnownTypesEnum.SHORT -> CodeBlock.of(
            "_gen.writeNumber(it)\n"
        )

        BINARY -> CodeBlock.of("_gen.writeBinary(it)\n")
        UUID -> CodeBlock.of("_gen.writeString(it.toString())\n")
    }
}
